### PR TITLE
FEAT: added dropdown with sign-out option in navbar

### LINKED ISSUE #284: Sign Out option in Navbar
## Issue
- User unable to sign out.

### Description
Currently, When the user visits the rds website, they sign in with git hub but are unable to sign out. This feature will enable them to sign out by providing them with an option to sign out in a dropdown from the navbar.

## Solution being implemented
- An API for signout in the backend 
    - PR for backend: https://github.com/Real-Dev-Squad/website-backend/pull/679
    - Backend ticket: https://github.com/Real-Dev-Squad/website-backend/issues/683

- We require An option to call for signout API in the front end
    - In navbar's user greeting we would add a dropdown that would contain links to sign out and visit your profile.

### PR DESCRIPTION
**Pull request by kotesh-arya**
Closes #284 

**Description** 
- Currently, When the user visits the RDS website, they sign in with Git hub but are unable to sign out. This feature will enable them to sign out by providing them with an option to sign out in a dropdown from the navbar, as shown below:
Screenshot (203)

### CODE REVIEW

File: components/dropdown/index.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+PATHS;
+const Dropdown = () => {
+  const signout = () => {
+    fetch('https://api.realdevsquad.com/auth/signout', {
+      method: 'GET',
+      credentials: 'include',
+    }).then(() => {
+      location.reload();
+    });
+    console.log(location);
```
Comment (vinayak-trivedi) [inline components/dropdown/index.js]: can you please remove this console.log

File: components/dropdown/index.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  };
+  return (
+    <div className={`${s.dropdown}`}>
```
Comment (vinayak-trivedi) [inline components/dropdown/index.js]: This will work I guess and can you please use camelCase everywhere so you will not need the brackets to access the classNames
```suggestion
    <div className={s.dropdown}>
```

File: components/dropdown/index.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+            <p className={`${s['profile-link']}`}>My profile</p>
+          </Link>
+        </li>
+        <hr className={s.line} />
+        <li
+          onClick={() => {
+            signout();
+          }}
+          className={`${s['dropdown-item']}`}
+          id={s['signout-option']}
```
Comment (vinayak-trivedi) [inline components/dropdown/index.js]: why are we using this Id?

File: components/dropdown/s.module.css [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  padding: 1rem;
+  text-align: left;
+  font-weight: 100;
+}
+
+.line {
+  border: none;
+  border-bottom: 2px solid black;
+  margin: 0;
+}
```
Comment (vinayak-trivedi) [inline components/dropdown/s.module.css]: can you put all the colors in css variables

**Review by vinayak-trivedi — CHANGES_REQUESTED** [review]
Thanks for the PR @kotesh-arya 
I've requested some changes, please have a look at them

File: components/dropdown/s.module.css (hunk 4 above)
Comment (kotesh-arya) [reply components/dropdown/s.module.css]: Sure, I will Implement all these requested changes 👍🏻

File: components/dropdown/index.js (hunk 3 above)
Comment (kotesh-arya) [reply components/dropdown/index.js]: I have used it to target the second option in the dropdown, Now I have updated it with a common class 👍🏻

File: components/dropdown/index.js (hunk 2 above)
Comment (kotesh-arya) [reply components/dropdown/index.js]: Thanks for pointing out, I have updated the classNames to camelCase

File: components/dropdown/index.js (hunk 1 above)
Comment (kotesh-arya) [reply components/dropdown/index.js:15]: Removed 👍🏻

**Review by Muralidhar22 — COMMENTED** [review]
I think the dropdown css file name can be given as dropdown.module.css, is there any reason it's named as s.module.css . what do you think?

File: components/dropdown/index.js (hunk 2 above)
Comment (MehulKChaudhari) [reply components/dropdown/index.js]: Please rename `s` to something meaningful

Comment (kotesh-arya) [reply components/dropdown/index.js]: In the latest commit I have updated the modular-styling to Styled-JSX for consistent styling. Will start naming files more meaningfully 👍🏻

**Comment by kotesh-arya** [comment]
I agree with you @Muralidhar22 , but I am using the styled-jsx now and I will consider naming files more meaningfully 👍🏻

File: components/dropdown/s.module.css (hunk 4 above)
Comment (Muralidhar22) [reply components/dropdown/s.module.css]: once you made the changes you can close the conversation

File: components/dropdown/index.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+const Dropdown = () => {
+  const signout = () => {
+    fetch('https://api.realdevsquad.com/auth/signout', {
+      method: 'GET',
+      credentials: 'include',
+    }).then(() => {
+      location.reload();
+    });
+  };
```
Comment (vinayak-trivedi) [inline components/dropdown/index.js]: please put this function in the utils folder

File: components/dropdown/index.js [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+            padding: 1rem;
+            text-align: left;
+            font-weight: 100;
+          }
+          .signoutOption:hover {
+            background-color: ${colors.grey.light};
+            border-radius: 0 0 0.5rem 0.5rem;
+          }
+        `}
+      </style>
```
Comment (vinayak-trivedi) [inline components/dropdown/index.js]: why are you putting the styles in the same file?

Comment (kotesh-arya) [reply components/dropdown/index.js]: I am Following the styled-JSX for consistency with the styles written for this project earlier

File: components/NavBar/index.js [hunk 7]
```diff
@@ -151,25 +153,26 @@ const NavBar = () => {
...
-              </Link>
+              <div className={styles.userGreetMsg}>
+                {`Hello ${isLoggedIn ? `${userData.firstName}` : 'User'}!`}
+              </div>
+              <img
+                className={styles.userProfilePic}
+                src={
+                  isLoggedIn
+                    ? `${userData.profilePicture}`
+                    : `${DEFAULT_AVATAR}`
```
Comment (vinayak-trivedi) [inline components/NavBar/index.js]: you are just putting the default avatar if the user is not logged in, but what will happen if the user is logged in and there is no profile URL in the user object?

File: package.json [hunk 8]
```diff
@@ -77,6 +77,7 @@
   },
   "volta": {
     "node": "14.19.1",
-    "npm": "6.14.16"
+    "npm": "6.14.16",
+    "yarn": "1.22.19"
```
Comment (vinayak-trivedi) [inline package.json]: why is this here?

File: components/NavBar/index.js (hunk 7 above)
Comment (kotesh-arya) [reply components/NavBar/index.js]: I have handled that case by showing up the Default avatar, if the logged-in user do not have a profile picture 👍🏻

File: color/color.json [hunk 9]
```diff
@@ -14,5 +16,10 @@
     "light": "#e4f7df",
     "primary": "#88d870",
     "dark": "#13380a"
+  },
```
Comment (Maheima) [inline color/color.json:16]: remove the extra spaces

**Review by Maheima — CHANGES_REQUESTED** [review]

File: color/color.json (hunk 9 above)
Comment (kotesh-arya) [reply color/color.json:16]: Sure I will 👍🏻

File: package.json (hunk 8 above)
Comment (kotesh-arya) [reply package.json]: Initially I used the yarn as the package manager, now I will be using npm and I have removed the un-used yarn.

File: color/color.json [hunk 10]
```diff
@@ -1,3 +1,5 @@
+
+
```
Comment (vinayak-trivedi) [inline color/color.json]: please remove these extra spaces

File: components/NavBar/index.js [hunk 11]
```diff
@@ -151,25 +153,30 @@ const NavBar = () => {
                 </button>
               </a>
             </Link>
+
```
Comment (vinayak-trivedi) [inline components/NavBar/index.js]: do we need this extra space?

File: components/dropdown/index.js [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+            cursor: pointer;
+          }
+
+          .signoutOption {
+            padding: 1rem;
+            text-align: left;
+            font-weight: 100;
+          }
+          .signoutOption:hover {
+            background-color: ${colors.grey.light};
```
Comment (vinayak-trivedi) [inline components/dropdown/index.js]: I don't think we need to put it inside curly braces, just check once, remove this in other places as well if that is the case

Comment (kotesh-arya) [reply components/dropdown/index.js]: Sure 👍🏻

File: components/NavBar/index.js (hunk 11 above)
Comment (kotesh-arya) [reply components/NavBar/index.js]: Will remove it in the upcoming commit 👍🏻

File: components/dropdown/index.js (hunk 12 above)
Comment (kotesh-arya) [reply components/dropdown/index.js]: Actually, here we are importing the colors object from the **color.json** , hence we are accessing it using curly braces only

Comment (vinayak-trivedi) [reply components/dropdown/index.js]: oh got it

File: components/dropdown/index.js [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+      <ul className="dropdownList">
+        <li className="dropdownItem">
+          <Link href={PATHS.PROFILE}>
+            <p className="profileLink">My Profile</p>
+          </Link>
+        </li>
+        <hr className="line" />
+        <li
+          onClick={() => {
+            signOut();
```
Comment (SanketDhabarde) [inline components/dropdown/index.js:19]: Why do we need a callback here? It can be like `onClick={signOut}`

File: components/dropdown/index.js [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+          }}
```
Comment (SanketDhabarde) [inline components/dropdown/index.js]: Why do we need a callback here? It can be like `onClick={signOut}`

File: components/NavBar/navbar.module.css [hunk 15]
```diff
@@ -65,6 +65,10 @@
 .userGreet {
   color: #ffffff;
   margin: 23px;
+  cursor: pointer;
+}
+.userGreet:hover {
+  color: #49a82e;
```
Comment (SanketDhabarde) [inline components/NavBar/navbar.module.css:71]: You can use variables for this color as well

Comment (kotesh-arya) [reply components/NavBar/navbar.module.css:71]: Sure I will use 👍🏻

File: components/dropdown/index.js (hunk 14 above)
Comment (kotesh-arya) [reply components/dropdown/index.js]: That's a better way 💯 , Will update. Thanks for pointing out 👍🏻

**Comment by Maheima** [comment]
I think you should add the functionality where the dropdown closes when a user clicks outside of it too

**Comment by MehulKChaudhari** [comment]
> I think you should add the functionality where the dropdown closes when a user clicks outside of it too

+1

**Comment by kotesh-arya** [comment]
> I think you should add the functionality where the dropdown closes when a user clicks outside of it too




Added it @Maheima  👍🏻

File: components/NavBar/index.js [hunk 16]
```diff
@@ -151,25 +158,30 @@ const NavBar = () => {
                 </button>
               </a>
             </Link>
+            {isLoggedIn && dropdownVisiblity && <Dropdown />}
             <div
+              onClick={(e) => {
+                e.stopPropagation();
+                setDropdownVisibility(!dropdownVisiblity);
+              }}
```
Comment (Muralidhar22) [inline components/NavBar/index.js]: why event propogation is added over here?

Comment (kotesh-arya) [reply components/NavBar/index.js]: I wanted to achieve the feature of closing the dropdown when user clicks outside the profile-name, In that case if user clicks on the user-name(the child element), then it's event is triggered and simultaneously it's parent's(the navbar) event also getting triggered. So in order to stop that propagation of event, I used **e.stopPropagation()**, inside the callback function of the event listener of the child element(user-name), from whom the event starts propagating.

Comment (Muralidhar22) [reply components/NavBar/index.js]: okay

**Review by Muralidhar22 — DISMISSED** [review]

**Comment by SanketDhabarde** [comment]
> > I think you should add the functionality where the dropdown closes when a user clicks outside of it too
> 
> Added it @Maheima 👍🏻

So @kotesh-arya will it close the dropdown when clicking on the navbar or anywhere on screen?

**Review by Maheima — DISMISSED** [review]
LGTM!

**Review by vinayak-trivedi — DISMISSED** [review]
LGTM 👍

**Comment by kotesh-arya** [comment]
> > > I think you should add the functionality where the dropdown closes when a user clicks outside of it too
> > 
> > 
> > Added it @Maheima 👍🏻
> 
> So @kotesh-arya will it close the dropdown when clicking on the navbar or anywhere on screen?

Yes @SanketDhabarde !
It will close even if user clicks anywhere on the page 👍🏻

File: redux/reducers/dropdowntoggle.js [hunk 17]
```diff
@@ -0,0 +1,21 @@
+import { CLOSE_DROPDOWN, OPEN_DROPDOWN } from '../actionTypes';
+
+const initialState = {
+  dropdownIsOpened: false,
+};
+
+export default function dropdownToggle(state = initialState, action) {
```
Comment (SanketDhabarde) [inline redux/reducers/dropdowntoggle.js:7]: Can you make a custom hook for this logic? (Clicking outside will close the component) So that it will be reusable for other modal components as well. Let me know if you want some reference

File: components/NavBar/index.js [hunk 18]
```diff
@@ -151,25 +161,32 @@ const NavBar = () => {
                 </button>
               </a>
             </Link>
+            {isLoggedIn && dropdownIsOpened && <Dropdown />}
             <div
+              onClick={(e) => {
+                e.stopPropagation();
+                dropdownIsOpened
+                  ? dispatch({ type: CLOSE_DROPDOWN })
+                  : dispatch({ type: OPEN_DROPDOWN });
```
Comment (MehulKChaudhari) [inline components/NavBar/index.js]: Can we please move it to a function and then use it? what will be easy to understand and increase readability.

File: components/dropdown/index.js [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+            cursor: pointer;
+          }
+
+          .signoutOption {
+            padding: 1rem;
+            text-align: left;
+            font-weight: 100;
+          }
+          .signoutOption:hover {
+            background-color: ${colors.grey.light};
```
Comment (MehulKChaudhari) [inline components/dropdown/index.js]: Using `module.css` instead of `style jsx` will be good?

**Review by MehulKChaudhari — CHANGES_REQUESTED** [review]
Great move 🔥 using redux. Added few comments please check.

**Review by Maheima — DISMISSED** [review]
LGTM!

File: redux/reducers/dropdowntoggle.js (hunk 17 above)
Comment (kotesh-arya) [reply redux/reducers/dropdowntoggle.js:7]: Sure I will👍🏻

File: components/NavBar/index.js (hunk 18 above)
Comment (kotesh-arya) [reply components/NavBar/index.js]: Moved the logic to a dedicated function 👍🏻

File: components/dropdown/index.js (hunk 19 above)
Comment (kotesh-arya) [reply components/dropdown/index.js]: I have moved this style-properties into a modular css file 👍🏻

File: redux/reducers/dropdowntoggle.js (hunk 17 above)
Comment (SanketDhabarde) [reply redux/reducers/dropdowntoggle.js:7]: https://usehooks.com/useOnClickOutside/

**Review by ivinayakg — APPROVED** [review]
Great Job 👍

**Review by MehulKChaudhari — APPROVED** [review]
